src/analysis/souffle/sensitivity_analysis.dl

```diff
@@ -0,0 +1,134 @@
+
+intermediateFlatten(list, sublist, list_element) :- isOperation(operation), 
+      operationHasOperandList(operation, list), 
+      list = [list_element, sublist].
+intermediateFlatten(list, sublist, list_element) :- intermediateFlatten(list, [list_element, sublist], _).
+
+.decl flatten(list:OperandList, list_element:AccessPath)
+flatten(list, list_element) :- intermediateFlatten(list, _, list_element).
+
+// Flattening each individual input per value. 
+.decl isSqlInput(operation: Operation, sql_input: AccessPath)
+
+isSqlInput(operation, sql_input) :- isSqlOperation(operation), 
+      operationHasOperandList(operation, input_list), 
+      flatten(input_list, sql_input). 
+
+// Keeping track of sensitivity score per each input at index. 
+// This is used to calculate the maximum input sensitivity. 
+// It's decomposed into base and inductive cases because recursive 
+// aggregates are not supported yet for the compiler case in souffle
+.decl sqlOperationHasSensitivityScoreAtIndex(operation: Operation, sensitivity_score: number, index: number) 
+
+// Base cases (index 0).
+// Sesnitivity score for initial inputs is 1. 
+sqlOperationHasSensitivityScoreAtIndex(operation, 1, 0) :- isSqlOperation(operation), 
+      operationHasOperandAtIndex(operation, sql_input, 0),
+      isInitialInput(sql_input).
+
+// For all other inputs, we get the final sensitivity score. 
+sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, 0) :- isSqlOperation(operation), 
+      operationHasOperandAtIndex(operation, sql_input, 0),
+      sqlValueSensitivity(sql_input, sensitivity_score).      
+
+// Inductive cases.  
+// We keep track at the score at each index, and only update if the new input score is higher. 
+sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, index) :- 
+      isSqlOperation(operation), 
+      operationHasOperandAtIndex(operation, sql_input, index),
+      isSqlInput(operation, sql_input), 
+      ( 
+            sqlValueSensitivity(sql_input, input_sensitivity_score), 
+            sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, index-1), 
+            sensitivity_score >= input_sensitivity_score; 
+            sqlValueSensitivity(sql_input, sensitivity_score), 
+            sqlOperationHasSensitivityScoreAtIndex(operation, prev_sensitivity_score, index-1), 
+            sensitivity_score >= prev_sensitivity_score
+      ). 
+
+// Function sensitivity scores per each sql operator as defined in PINQ
+// TODO: add other cases 
+.decl sqlFunctionSensitivity(operator: Operator, sensitivity_score: number)
+
+sqlFunctionSensitivity(operator, 1) :- operator = SQL_FILTER_OPERATOR;
+      operator = SQL_WHERE_OPERATOR; 
+      operator = SQL_SELECT_OPERATOR;
+      operator = SQL_COLUMN_ACCESS_OPERATOR;
+      operator = SQL_OUTPUT_OPERATOR; 
+      isAggregationOperation(operation),
+      operationHasOperator(operation, operator).
```

Comment:
I'd express the constant cases as base facts like so:

```
sqlFunctionSensitivity(SQL_FILTER_OPERATOR, 1).
sqlFunctionSensitivity(SQL_WHERE_OPERATOR, 1).
...
```

I know this is more verbose, but using base facts (rules without a `:-`) when possible reduces the complexity of the datalog in a number of important ways (such as being trivially stratifiable) which makes them worth the additional characters.

Reply:
Done.